relayer/processor/path_processor_internal.go

```diff
@@ -590,40 +619,37 @@ func (pp *PathProcessor) assembleAndSendMessages(
 	if len(messages.packetMessages) == 0 && len(messages.connectionMessages) == 0 && len(messages.channelMessages) == 0 {
 		return nil
 	}
-	var outgoingMessages []provider.RelayerMessage
+	outgoingMessages := make(
+		[]provider.RelayerMessage,
+		0,
+		len(messages.packetMessages)+len(messages.connectionMessages)+len(messages.channelMessages),
+	)
```

Comment:
The trick to making the slice writes lockless is to properly size the array up front, and allow each goroutine to write to a dedicated index.

That pattern looks like:

```go
out := make([]T, len(xs) + len(ys)) // len(out) == cap(out) here
idx := 0

for _, x := range xs {
  wg.Add(1)
  go processX(&wg, &out[i])
  i++
}

for _, y := range ys {
  wg.Add(1)
  go processY(&wg, &out[i])
  i++
}
```

Then you are left with a slice which may have nil values, so you need to decide whether to modify or copy the slice to omit the nil values, or update the `SendMessages` method to handle nils gracefully.

Alternatively, you can replace the appends in a channel and a collector goroutine:

```go
out := make([]T, 0, len(xs) + len(ys)) // Zero-length start this time.
ch := make(chan T, 4) // Arbitrary presize.

for _, x := range xs {
  wg.Add(1)
  go processX(&wg, ch)
}

for _, y := range ys {
  wg.Add(1)
  go processY(&wg, ch)
}

done := make(chan struct{}) // Signal that collection has finished.
go func() {
  for t := range <-ch {
    out = append(out, t)
  }
  close(done)
}()

wg.Wait()
close(ch) // Signal to collector that no more items remain.
<-done // Wait for collector to finish.

SendMessages(out)
```

Reply:
I explored these solutions while building this out, but both of them require data copy in this scenario unless we wanted to modify the `SendMessages` provider method to take a chan instead of a slice. I'm curious what your perspective is on why a data copy is better than using an isolated mutex for writes?

Reply:
For the batch message sending, using a chan seems like a lot of overhead when the slice accessor is so simple, e.g.: 
`batchMsgs := msgs[batchStartIdx:i]`. Also, within the sdk, it expects a slice of messages, so we would still have to convert to a slice at some point.

Reply:
I ran the following benchmarks:

```go
const testSliceSize = 100000

type TestStruct struct {
	FieldA string
	FieldB string
	FieldC string
	FieldD string
	FieldE string
	FieldF string
}

func testStruct(i int) TestStruct {
	return TestStruct{
		FieldA: fmt.Sprintf("fieldA%d", i),
		FieldB: fmt.Sprintf("fieldB%d", i),
		FieldC: fmt.Sprintf("fieldC%d", i),
		FieldD: fmt.Sprintf("fieldD%d", i),
		FieldE: fmt.Sprintf("fieldE%d", i),
		FieldF: fmt.Sprintf("fieldF%d", i),
	}
}

func sliceOfTestStructs() []TestStruct {
	s := make([]TestStruct, 0, testSliceSize)
	for i := 0; i < testSliceSize; i++ {
		s = append(s, testStruct(i))
	}
	return s
}

func testStructOptional(t TestStruct, i int) *TestStruct {
	if i%3 == 0 {
		return nil
	}
	return &t
}

func BenchmarkConcurrentSliceWriteLock(b *testing.B) {
	s := sliceOfTestStructs()
	n := make([]TestStruct, 0, testSliceSize)
	var wg sync.WaitGroup
	var l sync.Mutex
	for i, t := range s {
		wg.Add(1)
		i := i
		t := t
		go func() {
			defer wg.Done()
			o := testStructOptional(t, i)
			if o == nil {
				return
			}
			l.Lock()
			defer l.Unlock()
			n = append(n, *o)
		}()
	}

	wg.Wait()

}

func BenchmarkConcurrentSliceWriteCopy(b *testing.B) {
	s := sliceOfTestStructs()
	n := make([]*TestStruct, testSliceSize)
	var wg sync.WaitGroup
	for i, t := range s {
		wg.Add(1)
		i := i
		t := t
		go func() {
			defer wg.Done()
			n[i] = testStructOptional(t, i)
		}()
	}

	wg.Wait()

	a := make([]TestStruct, 0, testSliceSize)
	for _, t := range n {
		if t == nil {
			continue
		}
		a = append(a, *t)
	}
}

func BenchmarkConcurrentSliceWriteChan(b *testing.B) {
	s := sliceOfTestStructs()
	n := make(chan TestStruct, testSliceSize)
	var wg sync.WaitGroup
	for i, t := range s {
		wg.Add(1)
		i := i
		t := t
		go func() {
			defer wg.Done()
			o := testStructOptional(t, i)
			if o == nil {
				return
			}
			n <- *o
		}()
	}

	a := make([]TestStruct, 0, testSliceSize)
	done := make(chan struct{})

	go func() { // collector
		for t := range n {
			a = append(a, t)
		}
		close(done)
	}()

	wg.Wait()
	close(n) // Signal to collector that no more items remain.
	<-done   // Wait for collector to finish.

}
```

These are the results:

```bash
goos: linux
goarch: amd64
cpu: AMD Ryzen 9 5950X 16-Core Processor            
BenchmarkConcurrentSliceWriteLock
BenchmarkConcurrentSliceWriteLock-32            1000000000               0.09369 ns/op
BenchmarkConcurrentSliceWriteCopy
BenchmarkConcurrentSliceWriteCopy-32            1000000000               0.1148 ns/op
BenchmarkConcurrentSliceWriteChan
BenchmarkConcurrentSliceWriteChan-32            1000000000               0.1048 ns/op
```

When I shrink the testSliceSize to a lower number such as `500`, differences are similar:

```bash
goos: linux
goarch: amd64
cpu: AMD Ryzen 9 5950X 16-Core Processor            
BenchmarkConcurrentSliceWriteLock
BenchmarkConcurrentSliceWriteLock-32            1000000000               0.0000287 ns/op
BenchmarkConcurrentSliceWriteCopy
BenchmarkConcurrentSliceWriteCopy-32            1000000000               0.0000488 ns/op
BenchmarkConcurrentSliceWriteChan
BenchmarkConcurrentSliceWriteChan-32            1000000000               0.0000525 ns/op
```

Reply:
These benchmarks are not written properly: https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

The proposed change away from mutexes is not to solve a performance issue; it is to produce maintainable code that clearly shows how data is coordinated. 

Channel preference over mutexes follows the first proverb of Go:

> Don't communicate by sharing memory, share memory by communicating.

Reply:
> These benchmarks are not written properly

Thanks, updated with the for loop of b.N.

To me, the `BenchmarkConcurrentSliceWriteLock` seems to be the most succint and readable, but I am willing to be convinced otherwise.

Reply:
It helped me to step away from this for a bit and come back to it.

I generally don't have any issue with a mutex that is scoped to a particular data structure, e.g. if you had a struct like

```go
type ConcurrentDoubleMap struct {
  mu sync.Mutex
  
  byInt map[int]string
  byString map[string]int
}

func Add(int, string) {}
func LookupInt(int) string {}
func LookupString(string) int {}
```

But I would not want to see the same behavior captured as `DoSomething(ctx, &mu, intMap, stringMap)` even though that could function the same as the cleaner `DoSomething(ctx, &doubleMap)`

Since the slice is used in the unexported `assembleMessage` method already, how about we compromise by strictly scoping the mutex into a new type?

```go
// outgoingMessages is a slice of relayer messages that can be
// appended to concurrently.
type outgoingMessages struct {
  mu sync.Mutex
  msgs []*provider.RelayerMessage
}

// Append acquires a lock on om's mutex and then appends msg.
// When there are no more possible concurrent calls to Append,
// it is safe to directly access om.msgs.
func (om *outgoingMessages) Append(msg *provider.RelayerMessage) {
  om.mu.Lock()
  defer om.mu.Unlock()
  om.msgs = append(om.msgs, msg)
}
```

The compiler should inline calls to Append for effectively zero overhead, and then there is no ambiguity about the narrow scope of the mutex usage.